sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -289,6 +292,11 @@ abstract class Optimizer(catalogManager: CatalogManager)
    */
   def earlyScanPushDownRules: Seq[Rule[LogicalPlan]] = Nil
 
+  /**
+   * Override to provide additional rules for rewriting plans for v2 data sources.
+   */
+  def v2SourceRewriteRules: Seq[Rule[LogicalPlan]] = Nil
```

Comment:
I'd be open for other names if we want to make it more generic. Something like `rewriteRules`?

Reply:
After thinking more, I am inclined to call it `rewriteRules` as there may be other cases not related to v2 sources.

Reply:
Done.